modules/broker-core/src/main/java/org/wso2/broker/core/BindingSet.java

```diff
@@ -47,15 +49,36 @@ public static BindingSet emptySet() {
         unfilteredQueueBindings = new ConcurrentHashMap<>();
     }
 
-    boolean add(Binding binding) {
-        FieldValue selectorValue = binding.getArgument(Binding.JMS_SELECTOR_ARGUMENT);
-        Map<Queue, Binding> queueBindingMap;
-        if (selectorValue != null && !selectorValue.getValue().toString().isEmpty()) {
-            queueBindingMap = filteredQueueBindings;
-        } else {
-            queueBindingMap = unfilteredQueueBindings;
+    boolean add(Binding binding) throws ValidationException {
+
+        Binding existingBinding = validateBinding(binding);
+
+        if (Objects.isNull(existingBinding)) {
+            Map<Queue, Binding> queueBindingMap;
+            FieldValue selectorValue = binding.getArgument(Binding.JMS_SELECTOR_ARGUMENT);
+            if (Objects.nonNull(selectorValue) && !selectorValue.getValue().toString().isEmpty()) {
+                queueBindingMap = filteredQueueBindings;
+            } else {
+                queueBindingMap = unfilteredQueueBindings;
+            }
+            queueBindingMap.put(binding.getQueue(), binding);
+            return true;
         }
-        return queueBindingMap.putIfAbsent(binding.getQueue(), binding) == null;
+        return false;
+
+    }
+
+    private Binding validateBinding(Binding binding) throws ValidationException {
+        Binding existingBinding = unfilteredQueueBindings.get(binding.getQueue());
+        if (Objects.isNull(existingBinding)) {
+            existingBinding = filteredQueueBindings.get(binding.getQueue());
+        }
+
+        if (Objects.nonNull(existingBinding) && !existingBinding.equals(binding)) {
+            throw new ValidationException("Similar binding with different arguments already exist.");
```

Comment:
Shall we document this behavior somewhere?

Reply:
👍